File: artifacts/product-reality/sprint-192/m07/verify-census.mjs
import assert from 'node:assert/strict';
import fs from 'node:fs';
import path from 'node:path';
import {createHash} from 'node:crypto';
const base='artifacts/product-reality/sprint-192/m07';
const read=file=>JSON.parse(fs.readFileSync(file,'utf8'));
const hash=bytes=>createHash('sha256').update(bytes).digest('hex');
const after=read(`${base}/component-census/report.json`);
const before=read('artifacts/product-reality/sprint-191/m05/component-census/report.json');
const m05=read('artifacts/product-reality/sprint-192/m05/component-census-final/report.json');
const known=read('artifacts/product-reality/sprint-192/m05/census-diff.json');
const key=row=>`${row.input.object}/${row.input.context}`;
const changed=[];
assert.equal(after.greenTotalSchemas,77);assert.equal(after.greenTotalCells,154);
for(const row of after.allRows) {
 const old=before.allRows.find(prior=>key(prior)===key(row)),reviewed=m05.allRows.find(prior=>key(prior)===key(row));assert(old&&reviewed);
 const schema=read(row.composition.path).schema;
 assert.deepEqual(schema,read(reviewed.composition.path).schema,`${key(row)} moved beyond m05`);
 assert(row.cells.every(cell=>cell.errors.length===0));
 if(JSON.stringify(schema)!==JSON.stringify(read(old.composition.path).schema)) {
  const attribution=known.rows.find(prior=>key(prior)===key(row));assert(attribution,`unattributed schema ${key(row)}`);
  changed.push({...attribution,beforeSha256:hash(JSON.stringify(read(old.composition.path).schema)),afterSha256:hash(JSON.stringify(schema))});
 }
}
assert.equal(changed.length,14);
const classes=Object.fromEntries([...new Set(changed.map(row=>row.class))].map(name=>[name,changed.filter(row=>row.class===name).length]));
fs.writeFileSync(`${base}/schema-movement.json`,JSON.stringify({head:after.head,baseline:before.head,changedSchemas:changed.length,unchangedSchemas:77-changed.length,classes,rows:changed,unattributedChanges:[]},null,2)+'\n');
const original=read(`${base}/saved-original/report.json`),successor=read(`${base}/saved-successor/report.json`);
const hashes=report=>Object.fromEntries(report.rows.map(row=>[row.schema+'.json',row.input.sha256]));
assert.deepEqual(hashes(original),hashes(read('artifacts/product-reality/sprint-186/m05/reachability/report.json')));
const adoption=read('artifacts/product-reality/sprint-188/m01/adoption.json');assert.deepEqual(hashes(successor),Object.fromEntries(Object.entries(adoption.after).filter(([file])=>file!=='_index.json')));
assert.equal(original.reachable,15);assert.equal(successor.reachable,16);
const liveRoot='/Users/systemsystems/portfolio/Design-Tools/OODS-Forge/packages/mcp-server/.oods/schemas';
const live=Object.fromEntries(fs.readdirSync(liveRoot).filter(file=>file.endsWith('.json')).sort().map(file=>[file,hash(fs.readFileSync(path.join(liveRoot,file)))]));assert.deepEqual(live,adoption.after);
fs.writeFileSync(`${base}/saved-compatibility.json`,JSON.stringify({sourceHead:after.head,originalInputsUnchanged:true,successorInputsUnchanged:true,liveStoreFiles:Object.keys(live).length,changedLiveFiles:0,liveStore:{path:liveRoot,hashes:live},readOnly:true},null,2)+'\n');
console.log('77/77 schemas,154/154 generation cells;14 attributed schema movements; original15/16,successor16/16,17 unchanged live hashes.');

File: artifacts/product-reality/sprint-193/m07/verify-census.py
import json, collections, hashlib, subprocess
from pathlib import Path
base=Path('artifacts/product-reality/sprint-193/m07')
read=lambda file:json.loads(Path(file).read_text())
sha=lambda data:hashlib.sha256(data).hexdigest()
census=read(base/'component-census/report.json'); before=read('artifacts/product-reality/sprint-192/m07/component-census/report.json')
classes={r['id']:r['proposedClassification'] for r in read('packages/component-contracts/registry/component-reconciliation.proposed.v2.json')['rows']}
def nodes(n): return [n]+[v for c in n.get('children',[]) for v in nodes(c)]
def normalized(schema):
 ids={n['id']:f'node-{i}' for i,n in enumerate(n for s in schema['screens'] for n in nodes(s))}
 def clean(v):
  if isinstance(v,dict): return {k:clean(x) for k,x in v.items()}
  if isinstance(v,list): return [clean(x) for x in v]
  return ids.get(v,v) if isinstance(v,str) else v
 return json.dumps(clean(schema),sort_keys=True,separators=(',',':')).encode()
assert census['greenTotalSchemas']==77 and census['greenTotalCells']==154
attributions=read(base/'schema-attributions.json');rows=[]
for row in census['allRows']:
 old=next(r for r in before['allRows'] if r['input']==row['input']);a=read(old['composition']['path'])['schema'];b=read(row['composition']['path'])['schema'];key=row['input']['object']+'/'+row['input']['context']
 assert all(not c['errors'] for c in row['cells'])
 if normalized(a)==normalized(b): assert key not in attributions;continue
 attribution=attributions[key];assert attribution['reason'] and attribution['class']
 assert sha(normalized(a))==attribution['beforeNormalizedHash'] and sha(normalized(b))==attribution['afterNormalizedHash'],key
 ac=collections.Counter(n['component'] for s in a['screens'] for n in nodes(s));bc=collections.Counter(n['component'] for s in b['screens'] for n in nodes(s))
 items=lambda counter:[{'component':k,'count':v,'proposedClassification':classes[k]} for k,v in sorted(counter.items())]
 rows.append({'input':row['input'],'before':old['composition']['path'],'after':row['composition']['path'],**attribution,'added':items(bc-ac),'removed':items(ac-bc)})
assert len(rows)==len(attributions)
movement={'head':census['head'],'baseline':'c098237f1a1d026df4f1ad5c0ca51b15ebab0f4d','baselineReceiptHead':before['head'],'changedSchemas':len(rows),'unchangedSchemas':77-len(rows),'changedRuntimeCells':len(rows)*2,'classes':dict(collections.Counter(r['class'] for r in rows)),'rows':rows,'unattributedChanges':[],'normalization':'Only generated UI node IDs and exact references become deterministic traversal indices; every other schema byte participates. Component classes are proposals, not approvals.'}
(base/'schema-movement.json').write_text(json.dumps(movement,indent=2)+'\n')
original=read(base/'saved-original/report.json');successor=read(base/'saved-successor/report.json');hashes=lambda r:{v['schema']+'.json':v['input']['sha256'] for v in r['rows']}
assert hashes(original)==hashes(read('artifacts/product-reality/sprint-186/m05/reachability/report.json'))
adoption=read('artifacts/product-reality/sprint-188/m01/adoption.json');assert hashes(successor)=={k:v for k,v in adoption['after'].items() if k!='_index.json'}
assert original['reachable']==15 and successor['reachable']==16
live=Path('/Users/systemsystems/portfolio/Design-Tools/OODS-Forge/packages/mcp-server/.oods/schemas');actual={p.name:sha(p.read_bytes()) for p in sorted(live.glob('*.json'))};assert actual==adoption['after']
(base/'saved-compatibility.json').write_text(json.dumps({'sourceHead':census['head'],'originalInputsUnchanged':True,'successorInputsUnchanged':True,'liveStoreFiles':17,'changedLiveFiles':0,'liveStore':{'path':str(live),'hashes':actual},'readOnly':True},indent=2)+'\n')
print(f'77 schemas/154 generation cells; {len(rows)} attributed changes; original15/16, successor16/16,17 unchanged live hashes.')
